### PR TITLE
Include index prep

### PR DESCRIPTION
**Pull request by AlexAxthelm**
Add index preparation to steps of private docker image build.

Keeps pacta code and data in sync with what is included between index prep and CTM image

Depends on https://github.com/RMI-PACTA/workflow.prepare.pacta.indices/pull/97

### CODE REVIEW

**Comment by cjyetman** [comment]
I'll approve this once I have time to review it, but have you considered these impacts?
1 - Because the indices prep workflow includes data scraping, it's a bit more fragile and likely to fail than the rest of the CI here, and by including it here by default means this repo inherits that.
2 - Also wrt the data scraping aspect, I/we don't know how long those URLs will work as is, especially for older timestamps, so I feel a bit nervous that at some point the scraping for data from 2022.12.31 will stop working and then our CI here will fail.
3 - I suppose they don't run for too long, but I imagine this significantly increases the time for CI here to run?

File: .github/workflows/build-push-private.yml
Comment (cjyetman) [inline .github/workflows/build-push-private.yml:54]:
```suggestion
      templates-ref: ${{ steps.prepare.outputs.templates-ref }}
```

**Review by cjyetman — DISMISSED** [review]
I'm still a little confused how this works... we used to be quite strict about running the indices through the exact same PACTA process (which unfortunately creates that circular dependency), but since the indices processing is now part of the building process, it seems like the best possible situation is that the indices are processed with last previous commit on main, which will always be one commit before the commit the process is running for, no?

**Comment by jdhoffa** [comment]
> I'll approve this once I have time to review it, but have you considered these impacts? 1 - Because the indices prep workflow includes data scraping, it's a bit more fragile and likely to fail than the rest of the CI here, and by including it here by default means this repo inherits that. 2 - Also wrt the data scraping aspect, I/we don't know how long those URLs will work as is, especially for older timestamps, so I feel a bit nervous that at some point the scraping for data from 2022.12.31 will stop working and then our CI here will fail. 3 - I suppose they don't run for too long, but I imagine this significantly increases the time for CI here to run?

1 - Agree. In general, it's not the actual "scraping" that's important to do on every push to this repo, but rather the "running of PACTA". The intention is to keep the indices running in line with the state of the PACTA docker image, so that any updates to the code get automatically picked up into the data. One solution there could be to have indices scraping run in a different `workflow` and on some cadence (monthly? i don't know), and then the workflow triggered here would just pick up the latest scraped indices and "run PACTA on them" 
2 - See above (doesn't necessarily solve that, but at least extracts the "scraping" component from this particular repo) 
3 - It does indeed. One option would be to exclude the indices prep checks from the "required checks" before merging, though I'm not sure if that's desirable. Another option could be to run this on a CRON instead of on every PR, though I also don't know if that's desirable, because i guess we really do want to make sure that indices are up-to-date with the latest version of the code... all food for thought
Screenshot 2024-04-30 at 09 06 01

**Comment by cjyetman** [comment]
> I'm still a little confused how this works... we used to be quite strict about running the indices through the exact same PACTA process (which unfortunately creates that circular dependency), but since the indices processing is now part of the building process, it seems like the best possible situation is that the indices are processed with last previous commit on main, which will always be one commit before the commit the process is running for, no?

@jdhoffa does this not negate the primary purpose of this PR, to keep things "in sync"?

**Comment by jdhoffa** [comment]
> I'm still a little confused how this works... we used to be quite strict about running the indices through the exact same PACTA process (which unfortunately creates that circular dependency), but since the indices processing is now part of the building process, it seems like the best possible situation is that the indices are processed with last previous commit on main, which will always be one commit before the commit the process is running for, no?

I the indices would be prepared from MANY different categories of build:
* An intermediate build using the PR branch from which the action is triggered (this would be in line with whatever is on that branch)
* On a schedule
* On push to `main` (which of course would be the latest build from the latest commit on `main`)

Screenshot 2024-04-30 at 09 09 45

Screenshot 2024-04-30 at 09 11 59

**Comment by jdhoffa** [comment]
> Even on a run triggered by a push to main, seems like the `prepare-indices` job would have to run _after_ the `docker-build` job for that to be true.

Ah yeah I see the point, you're totally correct.

**Comment by cjyetman** [comment]
> > Even on a run triggered by a push to main, seems like the `prepare-indices` job would have to run _after_ the `docker-build` job for that to be true.
> 
> Ah yeah I see the point, you're totally correct.

to be clear, as I understand it, the input data would always be the same (which is arguably the most important bit), like even if the PR changes the target input data set in the config, but the PACTA code built-in to Docker that the indices are run through would always be one commit behind

**Review by jdhoffa — APPROVED** [review]

**Comment by jdhoffa** [comment]
> > Even on a run triggered by a push to main, seems like the `prepare-indices` job would have to run _after_ the `docker-build` job for that to be true.
> 
> Ah yeah I see the point, you're totally correct.

Hmm actually I think that is not the case. 
* The docker image is built on the latest commit using the `build_docker_image` action
* Then, the `build_private_docker_image` action pulls that most recent build from the registry, pulls the pacta-data, and runs indices from that most recently built image

In that sense, this is all run on the most recent build after all, and I believe this is good to go

Closing #305 without action and merging this

**Review by cjyetman — APPROVED** [review]
ok... I think @jdhoffa and @cjyetman have come to the conclusion that this does work as expected, because the Docker image without data is first created (including the latest version of all the code), and after that the indices are run through that freshly created Docker image, and the private Docker image with the data is built from the freshly created public Docker image, so everything is aligned in the end. A bit hard to follow, but seems like it works.

**Comment by AlexAxthelm** [comment]
> ok... I think @jdhoffa and @cjyetman have come to the conclusion that this does work as expected, because the Docker image without data is first created (including the latest version of all the code), and after that the indices are run through that freshly created Docker image, and the private Docker image with the data is built from the freshly created public Docker image, so everything is aligned in the end. A bit hard to follow, but seems like it works.

Sorry if didn't make this clear before I went out on PTO. Just to summarize the process overall

1. Push to `main` or PR with new code or build config
2. That triggers the build and push to ghcr of the "empty" (no data no private report templates) docker image using the code and build config
3. Run index prep using the newly created ghcr image and the pacta-data specified in the build config, and upload to Azure File Share
4. Run the private docker build action using the same `Dockerfile`, pacta-data, and the newly created index benchmarks, and push to Azure Container Registry

**Pull request merged**
